### PR TITLE
Listeners can log failures and shrinking attempts

### PR DESCRIPTION
**Pull request by giorgiosironi**
Solves https://trello.com/c/3OeQFdkx/32-an-user-can-run-a-test-and-see-what-was-the-first-failed-example-and-the-following-shrinking-process

### CODE REVIEW

**Comment by giorgiosironi** [comment]
@localheinz take a look at this PR, so that if I can get it in I can then merge your #76 and #83 without fears of conflict.

File: src/Eris/Listener.php [hunk 1]
```diff
@@ -17,9 +18,22 @@ public function startPropertyVerification();
...
+     * @param array  of mixed values
      * @param integer  index of current iteration
      * @return void
      */
-    public function newGeneration(array $generatedValues, $iteration);
+    public function newGeneration(array $generation, $iteration);
+
+    /**
+     * @param array  of mixed values
+     * @param Exception  assertion failure
```
Comment (localheinz) [inline src/Eris/Listener.php]: @giorgiosironi 

Should probably be

```php
/**
* @param array $generation array of mixed values
* @param Exception $exception  assertion failure
* @return void
 */
```

File: src/Eris/Listener.php [hunk 2]
```diff
@@ -17,9 +18,22 @@ public function startPropertyVerification();
...
+     * @return void
+     */
+    public function failure(array $generation, Exception $exception);
+
+    /**
+     * @param array  of mixed values
```
Comment (localheinz) [inline src/Eris/Listener.php]: @giorgiosironi 

Should probably be

```php
/**
 * @param array $generation array of mixed values
 * @return void
 */
```

File: src/Eris/Listener/Log.php [hunk 3]
```diff
@@ -25,16 +27,46 @@ public function __construct($file, $time, $pid)
...
-    public function newGeneration(array $generatedValues, $iteration)
+    public function newGeneration(array $generation, $iteration)
     {
-        $this->log("iteration $iteration");
+        $this->log(sprintf(
+            "iteration %d: %s",
+            $iteration,
+            // TODO: duplication with collect
+            json_encode(
+                $generation
```
Comment (localheinz) [inline src/Eris/Listener/Log.php]: @giorgiosironi 

Can we put this on one line instead?

```php
$this->log(sprintf(
    "iteration %d: %s",
    $iteration,
    // TODO: duplication with collect
    json_encode($generation)
));
```

File: src/Eris/Listener/Log.php [hunk 4]
```diff
@@ -25,16 +27,46 @@ public function __construct($file, $time, $pid)
...
         fclose($this->fp);
     }
 
+    public function failure(array $generation, Exception $exception)
+    {
+        $this->log(sprintf(
+            "failure: %s. %s",
+            // TODO: duplication with collect
+            json_encode(
+                $generation
```
Comment (localheinz) [inline src/Eris/Listener/Log.php]: @giorgiosironi 

Can we put this on one line instead?


```php
$this->log(sprintf(
    "failure: %s. %s",
    // TODO: duplication with collect
    json_encode($generation),
    $exception->getMessage()
));
```

File: src/Eris/Listener/Log.php [hunk 5]
```diff
@@ -25,16 +27,46 @@ public function __construct($file, $time, $pid)
...
+        ));
+    }
+
+    public function shrinking(array $generation)
+    {
+        $this->log(sprintf(
+            "shrinking: %s",
+            // TODO: duplication with collect
+            json_encode(
+                $generation
```
Comment (localheinz) [inline src/Eris/Listener/Log.php]: @giorgiosironi 

Can we put this on one line instead?


```php
$this->log(sprintf(
    "shrinking: %s",
    // TODO: duplication with collect
    json_encode($generation)
));
```

File: examples/LogFileTest.php [hunk 6]
```diff
@@ -17,5 +17,16 @@ public function testWritingIterationsOnALogFile()
                 $this->assertInternalType('integer', $number);
             });
     }
+
+    public function testLogOfFailuresAndShrinking()
+    {
+        $this->forAll(
+            Generator\int()
+        )
```
Comment (localheinz) [inline examples/LogFileTest.php]: @giorgiosironi 

Not sure, do you want to change the formatting here?

```php
$this
    ->forAll(Generator\int())
    ->hook(Listener\log('/tmp/eris-log-file-shrinking.log'))
    ->then(function($number) {
        $this->assertLessThanOrEqual(42, $number);
    })
;
```

File: docs/listeners.rst [hunk 7]
```diff
@@ -9,7 +9,11 @@ Consider that Eris performs (by default) 100 iterations for each ``forAll()`` in
 
 * ``startPropertyVerification()`` is called before the first iteration starts.
 * ``endPropertyVerification($ordinaryEvaluations)`` is called when no more iterations will be performed, both in the case of test success and failure. The ``$ordinaryEvaluations`` parameter provides the actual number of evaluations performed. This number may be less than than the number of iterations due to failures or ``when()`` filters not being satisfied.
-* ``newGeneration(array $generatedValues, $iteration)`` is called after generating a new iteration, and is passed the tuple of generated values which corresponds to the argument of ``then()`` along with the 0-based index of the iteration.
+* ``newGeneration(array $generation, $iteration)`` is called after generating a new iteration, and is passed the tuple of values along with the 0-based index of the iteration.
+* ``failure(array $generation, Exception $e)`` is called after the failure of an assertion (and not for generic exceptions). The method can be called only once per ``then()`` run.
```
Comment (localheinz) [inline docs/listeners.rst]: @giorgiosironi 

For consistency, do you want to use the same signature?

```diff
public function failure(array $generation, Exception $exception);
```

**Comment by localheinz** [comment]
@giorgiosironi 

Other than minor formatting issues which may be subject to personal taste, this looks good to me!

File: docs/listeners.rst (hunk 7 above)
Comment (giorgiosironi) [inline docs/listeners.rst]: you mean the variable names? That shouldn't a relevant part of the signature afaik

File: examples/LogFileTest.php (hunk 6 above)
Comment (giorgiosironi) [inline examples/LogFileTest.php]: Yes, I will

**Pull request merged**